File: bench/compare/go/vector_sort.go
package main

func siftDown(items []int64, start, end int) {
	root := start
	for {
		child := root*2 + 1
		if child >= end {
			return
		}
		candidate := root
		if items[candidate] < items[child] {
			candidate = child
		}
		right := child + 1
		if right < end && items[candidate] < items[right] {
			candidate = right
		}
		if candidate == root {
			return
		}
		items[root], items[candidate] = items[candidate], items[root]
		root = candidate
	}
}

func heapSort(items []int64) {
	for start := len(items) / 2; start > 0; {
		start--
		siftDown(items, start, len(items))
	}
	for end := len(items); end > 1; {
		end--
		items[0], items[end] = items[end], items[0]
		siftDown(items, 0, end)
	}
}

func binarySearch(items []int64, needle int64) int64 {
	lo := 0
	hi := len(items)
	for lo < hi {
		mid := lo + (hi-lo)/2
		value := items[mid]
		if value == needle {
			return int64(mid)
		}
		if value < needle {
			lo = mid + 1
		} else {
			hi = mid
		}
	}
	return -1
}

func buildReverse(n int64) []int64 {
	items := make([]int64, 0, n)
	for i := n - 1; i >= 0; i-- {
		items = append(items, i)
	}
	return items
}

func checksum(items []int64) int64 {
	var sum int64
	for _, value := range items {
		sum += value
	}
	return sum
}

func runOnce(n int64) int64 {
	items := buildReverse(n)
	heapSort(items)
	return checksum(items) + binarySearch(items, 0) + binarySearch(items, n/2) + binarySearch(items, n-1)
}

func main() {
	n := int64(600)
	runs := 5
	var total int64
	for i := 0; i < runs; i++ {
		total += runOnce(n)
	}
	if total != 902995 {
		panic(total)
	}
}

File: bench/compare/rust/vector_sort.rs
fn sift_down(items: &mut [i64], start: usize, end: usize) {
    let mut root = start;
    loop {
        let child = root * 2 + 1;
        if child >= end {
            return;
        }
        let mut candidate = root;
        if items[candidate] < items[child] {
            candidate = child;
        }
        let right = child + 1;
        if right < end && items[candidate] < items[right] {
            candidate = right;
        }
        if candidate == root {
            return;
        }
        items.swap(root, candidate);
        root = candidate;
    }
}

fn heap_sort(items: &mut [i64]) {
    let mut start = items.len() / 2;
    while start > 0 {
        start -= 1;
        sift_down(items, start, items.len());
    }
    let mut end = items.len();
    while end > 1 {
        end -= 1;
        items.swap(0, end);
        sift_down(items, 0, end);
    }
}

fn binary_search(items: &[i64], needle: i64) -> i64 {
    let mut lo = 0_usize;
    let mut hi = items.len();
    while lo < hi {
        let mid = lo + (hi - lo) / 2;
        let value = items[mid];
        if value == needle {
            return mid as i64;
        }
        if value < needle {
            lo = mid + 1;
        } else {
            hi = mid;
        }
    }
    -1
}

fn build_reverse(n: i64) -> Vec<i64> {
    let mut items = Vec::with_capacity(n as usize);
    let mut i = n - 1;
    loop {
        items.push(i);
        if i == 0 {
            break;
        }
        i -= 1;
    }
    items
}

fn checksum(items: &[i64]) -> i64 {
    items.iter().copied().sum()
}

fn run_once(n: i64) -> i64 {
    let mut items = build_reverse(n);
    heap_sort(&mut items);
    checksum(&items)
        + binary_search(&items, 0)
        + binary_search(&items, n / 2)
        + binary_search(&items, n - 1)
}

fn main() {
    let n = 600_i64;
    let runs = 5;
    let mut total = 0_i64;
    for _ in 0..runs {
        total += run_once(n);
    }
    if total != 902995 {
        panic!("{}", total);
    }
}
